rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,258 @@
+
+Large Vector users often require complex Vector topologies to facilitate the collection and processing of data from many different upstream sources. This results in large Vector configuration files that are hard to manage across different teams. This makes Vector a bad candidate for large teams that require autonomy to facilitate sane collaboration.
+
+## Context
+
+- [RFC 2064](https://github.com/timberio/vector/blob/master/rfcs/2020-03-17-2064-event-driven-observability.md) Event drive observability.
+- [RFC 8288 (PR)](https://github.com/timberio/vector/pull/8400) Adding the `resources` component.
+
+## Scope
+
+### In the scope
+
+- How and where the pipelines are defined
+- How vector reads the pipelines and what are the limitations of pipelines
+- How do we monitor pipelines components
+
+### Out of scope
+
+- How the pipelines should be synchronized between vector instances
+- Pre/post processing of data - The ability to prepare and normalize data.
+- Connecting pipelines together - The ability to take input from another pipeline.
+- Component reuse - The ability to define boilerplate for reuse across many different pipelines. This will likely align with Datadog’s “pipeline catalogue”.
+- Pipeline quotas - The ability to limit how much data a pipeline can send to a sink.
+- Visibility of the components - The ability to give access to any components within the system.
+
+## Pain
+
+- If an admin wants to split responsibilities between `ops` and `devs` by only allowing the `ops` to deal with `sources/sinks` and the `devs`, it's currently not possible.
+
+## User Experience
+
+These changes are providing a way for vector users to split their configuration in a way that improves the collaboration between ops and devs.
+This split will take be made by creating a pipeline configuration file, in which the devs will be able to configure their components.
+The ops will now configure their `sources` and `sinks`, and expose them to be used by the devs.
+The ops will be able to only given access to some of the common components when configuring the pipeline.
+The devs will have the ability to consume the provided components without being able to change the common configuration.
+The users will be able to monitor their pipelines through the `internal_metrics` source.
+
+## Motivation
+
+- Helps Vector to grow organically within an organization by allowing teams to adopt Vector at their own pace without heavy administrator involvement.
+- Reduces the management overhead of devops/SREs by enabling teams to manage their own pipelines (spread the management load).
+
+## Implementation
+
+### How and where the pipelines are stored
+
+To avoid backward incompatibility, pipelines will be loaded from a `pipelines` sub-directory relative to the Vector configuration directory (e.g., `/etc/vector/pipelines`). Therefore, if a user changes the location of the Vector configuration directory they will also change the `pipelines` directory path. They are coupled.
```

Comment:
This was my idea, but I've since backtracked on it after exploring a pattern of coupling pipeline configs with applications. Imagine checking in your service-specific pipeline config with your application repository and then delivering that config alongside your app. In this scenario pipeline configs could live in multiple places.

To be clear, the `pipelines` subdirectory makes sense as a default, but we should allow configuration to enable the above pattern.

Reply:
or even have multiple sources if I follow your logic? Have a folder dedicated per team for example...?

Reply:
After discussing with @jdrouet, we will follow up with this improvement.

Reply:
maybe we should do that in a future RFC, to keep this simple for now?

Reply:
I'd like to see this noted in "out of scope", as I could see a more flexible `pipelines` dir being a common request we'll want to circle back to.

Reply:
Yes, as such, I would like to refrain from exposing any sort of pipelines direction configuration option since we'll likely want to support multiple directories in the future.

@jdrouet can you note this in the "out of scope" section, ensure this RFC does not propose a pipelines directory configuration option, and then close this comment?